### PR TITLE
Updated code_utils.py & local_commandline_code_executor.py (powershell to pwsh)

### PR DESCRIPTION
**Pull request by abhaymathur21**
Updated the 'powershell' command to 'pwsh' for a ps1 executable file 

## Why are these changes needed?

I was trying to use autogen to execute terminal commands (eg: git status, ls) but it was throwing an error. So these changes are required if terminal commands are to be executed in the new 7.x version of PowerShell. To execute terminal commands in powershell, autogen first makes a ps1 shell script with those commands written in it and then executes that ps1 shell script. In order for that shell script execution to work, these changes are required.

In March 2020, microsoft released a new version of powershell (7.0) in which the powershell.exe file was renamed to pwsh.exe 
image
https://learn.microsoft.com/en-us/powershell/scripting/whats-new/differences-from-windows-powershell?view=powershell-7.4

In your code, if the user is working in a powershell window and a '.ps1' script is to be executed, your code uses the 'powershell' command to execute it which has now been replaced with 'pwsh' in the new Microsoft PowerShell version 7.0
image
image

So in your code I replaced the command 'powershell' with 'pwsh' when a ps1 script has to be executed

I also added the 'pwsh' command to the list of commands in the local_commandline_code_executor.py file 
image

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

### CODE REVIEW

File: autogen/code_utils.py
Comment (ekzhu) [inline autogen/code_utils.py:222]: Shouldn’t the first condition need to be split to also handle “powershell”?

**Review by ekzhu — CHANGES_REQUESTED** [review]
Thanks for the PR! Have you also checked out the new code executor? It’s here https://github.com/microsoft/autogen/blob/main/autogen/coding/local_commandline_code_executor.py 

You can use the new code executor like

```python
proxy = CoversableAgent(“proxy”, code_execution_config={“executor”: “commadline-local”})
```
Perhaps you want to add the handling for powershell there as well? It’s okay to do it in a separate PR.

File: autogen/code_utils.py
Comment (abhaymathur21) [reply autogen/code_utils.py:222]: yes youre right, thanks for pointing that out, i have added a split condition to deal with powershell as well could you please review that once as well?

**Comment by abhaymathur21** [comment]
@microsoft-github-policy-service agree

File: autogen/code_utils.py
Comment (abhaymathur21) [reply autogen/code_utils.py:222]: I added the pwsh command to the local commandline executor file list of commands as well, please review that once too

File: autogen/coding/local_commandline_code_executor.py
Comment (ekzhu) [inline autogen/coding/local_commandline_code_executor.py:131]:
```suggestion
            if lang in ["bash", "shell", "sh", "pwsh", "powershell", "ps1"]:
```

Comment (ekzhu) [reply autogen/coding/local_commandline_code_executor.py:131]: The `lang` here is only used for filename, which doesn't affect how it is executed. It uses `code_utils` later so the execution part is already handled.

Comment (abhaymathur21) [reply autogen/coding/local_commandline_code_executor.py]: oh right okay

**Review by ekzhu — APPROVED** [review]

File: autogen/code_utils.py
Comment (sonichi) [inline autogen/code_utils.py:223]: why not merge the conditions in line 221 and 223?
Can we add a simple test to cover the if branch?

Comment (abhaymathur21) [reply autogen/code_utils.py:223]: But then if lang = powershell we wont have any case to deal with that right? For the users that still use old powershell versions. Is there any other way we can cover that if branch with a test that I am unaware of at the moment?

Comment (sonichi) [reply autogen/code_utils.py:223]: I mean changing line 221 into: `if lang in ["ps1", "powershell"]:`

Comment (abhaymathur21) [reply autogen/code_utils.py:223]: Ohh okay got it yes you're right that would be more efficient, sorry I'll make the changes right away

**Comment by sonichi** [comment]
@ekzhu @abhaymathur21 have you tested it in windows? When I run pwsh in windows, it can't find the command. powershell is available. I think this change will break for existing users. We'd better check whether pwsh is available in the current system environment.

**Comment by abhaymathur21** [comment]
> @ekzhu @abhaymathur21 have you tested it in windows? When I run pwsh in windows, it can't find the command. powershell is available. I think this change will break for existing users. We'd better check whether pwsh is available in the current system environment.


image

I have tested it in windows yes. I was getting an error that's why I did some research and found out that since mine is a relatively new laptop, it came with the 7.0 PowerShell installed so the command that works for me is 'pwsh'. But for users that still operate 5.0 PowerShell the command is 'powershell'. So is there any way we can customize this command for the user in our code based on the version of powershell they are using maybe?

**Comment by sonichi** [comment]
I asked on Discord. I prefer detecting the availability or making it configurable.

**Comment by abhaymathur21** [comment]
> I asked on Discord. I prefer detecting the availability or making it configurable.

I added a function to detect the command (powershell or pwsh) which works in the user's system and then use that command only
image
image

Is this okay? I tested it, it works with autogen on my PC

**Comment by sonichi** [comment]
That works. Could you add a unit test to cover the new function?

**Comment by abhaymathur21** [comment]
> That works. Could you add a unit test to cover the new function?

Sure, I wrote this unit test and added it to your test_code_utils.py (which was incorrectly named as test_code.py initially but I corrected it to test_code_utils.py using my own discretion):

class TestGetPowerShellCommand(unittest.TestCase):
    
    @patch('subprocess.run')
    def test_get_powershell_command_powershell(self, mock_subprocess_run):
        # Set up the mock to return a successful result for 'powershell'
        mock_subprocess_run.return_value.returncode = 0
        mock_subprocess_run.return_value.stdout = StringIO("5")
        
        self.assertEqual(get_powershell_command(), "powershell")
    
    @patch('subprocess.run')
    def test_get_powershell_command_pwsh(self, mock_subprocess_run):
        # Set up the mock to return a successful result for 'pwsh'
        mock_subprocess_run.side_effect = [FileNotFoundError, mock_subprocess_run.return_value]
        mock_subprocess_run.return_value.returncode = 0
        mock_subprocess_run.return_value.stdout = StringIO("7")
        
        self.assertEqual(get_powershell_command(), "pwsh")
    
    @patch('subprocess.run')
    def test_get_powershell_command_no_shell(self, mock_subprocess_run):
        # Set up the mock to simulate 'powershell' and 'pwsh' not found
        mock_subprocess_run.side_effect = [FileNotFoundError, FileNotFoundError]
        
        with patch('sys.stdout', new=StringIO()) as fake_out:
            get_powershell_command()
            self.assertEqual(fake_out.getvalue().strip(), "Neither powershell nor pwsh is installed.")
    
    @patch('subprocess.run')
    def test_get_powershell_command_no_shell_no_output(self, mock_subprocess_run):
        # Set up the mock to simulate 'powershell' and 'pwsh' not found without printing error message
        mock_subprocess_run.side_effect = [FileNotFoundError, FileNotFoundError]
        
        self.assertIsNone(get_powershell_command())
        
if __name__ == '__main__':
    unittest.main()

**Comment by sonichi** [comment]
Thanks. That works. The conflict needs to be resolved though before merging.

**Comment by abhaymathur21** [comment]
> Thanks. That works. The conflict needs to be resolved though before merging.

Done!

File: autogen/code_utils.py
Comment (sonichi) [inline autogen/code_utils.py:244]:
```suggestion
    if lang in ["ps1", "pwsh", "powershell"]:
```
Let's add alternative strs here because the language model may still suggest them.

Comment (abhaymathur21) [reply autogen/code_utils.py]: oh yes thats a good idea

**Pull request merged**
